### ISSUE TITLE
Display config.log after build failures

### ISSUE THREAD

**Issue by chrissimpkins** [source-foundry/ttfautohint-build#1]
TODO:

- [X] find path to Harfbuzz console.log file
- [x] add `after_failure` scripts to .travis.yml file to display console.log for Freetype, Harfbuzz, ttfautohint

```
cat $HOME/ttfautohint-build/freetype-2.8/builds/unix/config.log
cat $HOME/ttfautohint-build/harfbuzz-1.5.0/config.log
cat $HOME/ttfautohint-build/ttfautohint-1.7/config.log
```

**Comment by chrissimpkins** [comment]
@lemzwerg 

Werner, we can continue the conversation about build failures on OS X over here.  Let me investigate where Harfbuzz drops the console.log files on builds and I should be able to add this to the travis script today.

**Comment by lemzwerg** [comment]
@chrissimpkins, it's rather the `config.log` file of ttfautohint that I need...

**Comment by chrissimpkins** [comment]
Is it worth displaying config.log of all three with every failure?

**Comment by lemzwerg** [comment]
the path to HarfBuzz's `config.log` file is exactly the same pattern:
```
$HOME/ttfautohint-build/harfbuzz-1.5.0/config.log
```

**Comment by lemzwerg** [comment]
I don't know how Travis works, so I can't give a recommendation.  The best solution (w.r.t. legibility) would be to emit only the 'current' config.log file in case of failure.

**Comment by chrissimpkins** [comment]
> The best solution (w.r.t. legibility) would be to emit only the current package's config.log file in case of failure.

Definitely agree.  We have the capacity to catch zero and non-zero exit status codes from the tests on Travis, then execute additional scripts accordingly.

We could approach this by developing a script that detects missing build product(s) from each of the dependency and ttfautohint build steps that is executed on any build failure (non-zero exit status code).  This would not need to be a part of the ttfautohint-build.sh script.  It can be a separate script in the repository that is only triggered on failed builds during Travis CI testing.  

Could be something along the lines of the following:

```shell
if ! [ -f "path/to/freetype/somefile ]; then
    echo "======  FREETYPE CONSOLE.LOG ========"
    cat path/to/freetype/console.log
fi
    
if ! [ -f "path/to/harfbuzz/somefile ]; then
    echo "======  HARFBUZZ CONSOLE.LOG ========"
    cat path/to/harfbuzz/console.log
fi

if ! [ -f "path/to/ttfautohint/somefile ]; then
    echo "======  TTFAUTOHINT CONSOLE.LOG ========"
    cat path/to/ttfautohint/console.log
fi

```

**Comment by chrissimpkins** [comment]
or better yet in order to avoid raising additional errors for previous build steps that prevented subsequent build attempts, we can incorporate the sequential build strategy:

```
if ! [ -f "path/to/freetype/somefile ]; then
    echo "======  FREETYPE CONFIG.LOG ========"
    cat path/to/freetype/config.log
else
    if ! [ -f "path/to/harfbuzz/somefile ]; then
        echo "======  HARFBUZZ CONFIG.LOG ========"
        cat path/to/harfbuzz/config.log
    else
        if ! [ -f "path/to/ttfautohint/somefile ]; then
            echo "======  TTFAUTOHINT CONFIG.LOG ========"
            cat path/to/ttfautohint/config.log
        fi
    fi
fi
    



```

**Comment by lemzwerg** [comment]
Alternatively, `ttfautohint-build.sh` could emit different exit codes, depending on the reached stage.

**Comment by chrissimpkins** [comment]
> Alternatively, ttfautohint-build.sh could emit different exit codes, depending on the reached stage.

That's a good idea.  Would need to modify the build script to capture failures and raise these exit status codes, then we could have something along the lines of:

##### logging.sh

```
if [ $? -eq 2 ]; then
    echo "======  FREETYPE CONFIG.LOG ========"
    cat path/to/freetype/config.log
else
    if [ $? -eq 3 ]; then
        echo "======  HARFBUZZ CONFIG.LOG ========"
        cat path/to/harfbuzz/config.log
    else
        if [ $? -eq 4 ]; then
            echo "======  TTFAUTOHINT CONFIG.LOG ========"
            cat path/to/ttfautohint/config.log
        fi
    fi
fi
```

**Comment by chrissimpkins** [comment]
I am not certain how Travis passes the exit status code on to the `after_failure` script execution.  It is possible that we don't see the exit status of the `ttfautohint-build.sh` script immediately preceding the log print scripts.

**Comment by lemzwerg** [comment]
BTW, you might use `elif` in shell scripts instead of `else if`.

**Comment by chrissimpkins** [comment]
👍 

```
if [ $? -eq 2 ]; then
    echo "======  FREETYPE CONFIG.LOG ========"
    cat path/to/freetype/config.log
elif [ $? -eq 3 ]; then
    echo "======  HARFBUZZ CONFIG.LOG ========"
    cat path/to/harfbuzz/config.log
elif [ $? -eq 4 ]; then
    echo "======  TTFAUTOHINT CONFIG.LOG ========"
    cat path/to/ttfautohint/config.log
fi
```

Do you want me to make changes or would you like to do this?

**Comment by lemzwerg** [comment]
Please proceed :-)  As soon as I have some time I will check everything.

**Comment by chrissimpkins** [comment]
Deal.  I will PR in changes from temporary working branches.

**Comment by chrissimpkins** [comment]
From Travis docs:

> Note on $?

>Each command in script is processed by a special bash function. This function manipulates $? to produce logs suitable for display. Consequently, you should not rely on the value of $? in script section to alter the build behavior. See this [GitHub issue](https://github.com/travis-ci/travis-ci/issues/3771) for a more technical discussion.

**Comment by chrissimpkins** [comment]
removed OS X 10.10 testing from master branch.  Will leave OS X 10.10 test failures in the branch 

https://github.com/source-foundry/ttfautohint-build/tree/osx10.10-testing

**Comment by lemzwerg** [comment]
... just wondering: Is there a possibility to quickly get ttfautohint's `config.log` file for OS X 10.10?  I would like to proceed with the gnulib bug report...

**Comment by chrissimpkins** [comment]
absolutely! sorry didn't mean that work wouldn't continue on this.  just branched work so that we have passing tests on what is in master.  

should be able to get to it today.

**Comment by chrissimpkins** [comment]
I was able to get the raw log, pulled the ttfautohint config.log text, and posted it here:

https://gist.github.com/chrissimpkins/c2461678c20f9fc532678a4e70f2dfed

**Comment by chrissimpkins** [comment]
For now, the workaround that I am using is to maintain individual branches for freetype, harfbuzz, and ttfautohint config.log prints to stdout on build failures.

Can push a commit to any branch to prompt a new build attempt and view the respective config.log on failure.

- `osx10.10-testing-freetype`
- `osx10.10-testing-harfbuzz`
- `osx10.10-testing-ttfautohint`

**Comment by chrissimpkins** [comment]
Discussion about identified build issue on OS X 10.10 is in the thread #5

**Comment by chrissimpkins** [comment]
Werner fixed OX 10.10 bug in #5.  This will be fixed in ttfautohint v1.7.1 release.  Closing this issue.
